### PR TITLE
[omni, tests] feat: add general image editing interface

### PR DESCRIPTION
**Pull request by NancyFyong**
## Summary

This PR adds the model-agnostic pieces needed by image-editing pipelines:

- `ImageGenerationRequest` normalizes text-to-image and image-conditioned rollout payloads.
- `DiffusionI2IModelBase` defines condition preparation, concat/crop injection, metadata unwrapping, and sequence-parallel validation.
- `prepare_processor_files()` runs before both driver-side and worker-side processor loading and may return a writable prepared path for read-only model mounts.
- CPU tests cover request parsing, T2I compatibility, I2I condition dispatch, processor load order, and external adapter overrides.

The branch was rebuilt from `upstream/main` at `f81ee27`. It contains no Qwen-Image-Edit, FLUX, reward-service, training-recipe, or long-form documentation changes. The Qwen implementation is isolated on `NancyFyong/pr/qwen-image-edit` and will be opened against `main` after this PR merges.

## Compatibility

Existing T2I adapters keep the original `prepare_model_inputs` path. I2I adapters opt in by subclassing `DiffusionI2IModelBase`; missing or malformed condition latents fail before model execution. Driver and worker processor resolution use the same registry and external-library override order.

## Duplicate-work check

Checked before rebuilding:

- `gh issue view 237 --repo verl-project/verl-omni --comments`
- `gh pr list --repo verl-project/verl-omni --state open --search "237 in:body"`
- `gh pr list --repo verl-project/verl-omni --state open --search "image editing interface in:title,body"`
- `gh pr list --repo verl-project/verl-omni --state open --search "Qwen Image Edit FlowGRPO in:title,body"`

No separate open PR implements this interface.

## Tests

- `PYTHONPATH=/tmp/kilo/verl-omni-general pytest -q tests/pipelines/test_image_edit_interface_on_cpu.py tests/pipelines/test_i2i_model_base_on_cpu.py tests/pipelines/test_diffusion_model_base_on_cpu.py tests/workers/config/test_diffusion_config_on_cpu.py` - 59 passed
- `ruff check` and `ruff format --check` on all changed Python files - passed
- full pre-commit hooks on every commit - passed
- `git diff --check upstream/main..HEAD` - passed

## AI assistance

AI assistance was used while restructuring and testing this change. I reviewed the final diff and test results.

### CODE REVIEW

File: verl_omni/workers/config/diffusion/model.py [hunk 1]
```diff
@@ -156,10 +156,18 @@ def __post_init__(self):
...
+                from verl_omni.pipelines.model_base import DiffusionModelBase
+
+                try:
+                    DiffusionModelBase.get_class(self).prepare_processor_files(self.local_path)
+                except NotImplementedError:
+                    logger.debug(
+                        "No diffusion model adapter registered for %s; skipping processor preparation hook.",
+                        self.architecture,
+                    )
+                self.processor = hf_processor(processor_path, trust_remote_code=self.trust_remote_code)
```
Comment (zhtmike) [inline verl_omni/workers/config/diffusion/model.py:169]: drop `except NotImplementedError:`, since it will not be reached.

File: verl_omni/pipelines/model_base.py [hunk 2]
```diff
@@ -102,6 +102,11 @@ def configure_train_mode(cls, module: torch.nn.Module) -> None:
         """Hook called after ``module.train()`` for architecture-specific overrides."""
         return
 
+    @classmethod
+    def prepare_processor_files(cls, model_path: str) -> None:
```
Comment (zhtmike) [inline verl_omni/pipelines/model_base.py]: what is the hook aimed for?

File: verl_omni/pipelines/image_edit_utils.py [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+        )
+
+
+@dataclass
+class ImageEditReplay:
+    """Replay state produced by image-editing rollouts and consumed by training.
+
+    Only the sampling trajectory is required and cross-model. Model-specific
+    conditioning (prompt/image embeddings, condition latents, guidance state) is
+    carried opaquely in ``condition`` so the contract stays backend-agnostic.
```
Comment (zhtmike) [inline verl_omni/pipelines/image_edit_utils.py]: pls add field docstring & illustrations, like `condition` and `extras`, what will be in such field.

Do you think we need such interface for t2i as well?

File: verl_omni/pipelines/image_edit_utils.py [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+    if metadata is None:
+        metadata = custom_prompt.get("extra_info")
+    if isinstance(metadata, Mapping):
+        return metadata
+
+    additional_information = custom_prompt.get("additional_information")
+    if isinstance(additional_information, Mapping):
+        return additional_information
+
+    return None
```
Comment (zhtmike) [inline verl_omni/pipelines/image_edit_utils.py]: it is too AI and overly complicated, I suggest just remove them

File: verl_omni/pipelines/image_edit_utils.py [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+        value = value[key]
+    return value
+
+
+def _normalize_images(images: Any) -> list[Any]:
+    if isinstance(images, list):
+        return images
+    if isinstance(images, tuple):
+        return list(images)
+    return [images]
```
Comment (zhtmike) [inline verl_omni/pipelines/image_edit_utils.py]: same

File: verl_omni/pipelines/image_edit_utils.py [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+        return cls(
+            all_latents=custom_output["all_latents"],
+            all_log_probs=custom_output["all_log_probs"],
+            all_timesteps=custom_output["all_timesteps"],
+            condition=custom_output.get("condition"),
+            extras=custom_output.get("extras"),
+        )
+
+
+def _resolve_image_edit_images(custom_prompt: Mapping[str, Any]) -> list[Any] | None:
```
Comment (SamitHuang) [inline verl_omni/pipelines/image_edit_utils.py]: if it's for image normaliztion mainly, we make the function name strateforward for clarity

File: verl_omni/pipelines/image_edit_utils.py [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+        ("extra_args", "multi_modal_data", "image"),
+        ("additional_information", "condition_images"),
+    ):
+        images = _get_nested(custom_prompt, path)
+        if images is not None:
+            return _normalize_images(images)
+    return None
+
+
+def _resolve_metadata(custom_prompt: Mapping[str, Any]) -> Mapping[str, Any] | None:
```
Comment (SamitHuang) [inline verl_omni/pipelines/image_edit_utils.py]: ditto. need a clearer func name

**Review by SamitHuang — COMMENTED** [review]
Thanks for the work. I think it'd be good to discuss these image edit API changes interactively on  community meeting, if you feel comfortable to join :)

File: verl_omni/pipelines/image_edit_utils.py [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+(the FlowGRPO family). Image-editing backends consume ImageEditRequest inputs
+from the existing DataProto/custom_prompt path and emit ImageEditReplay fields
+through DiffusionOutput.custom_output for training replay.
+
+The only cross-model required replay fields are the sampling trajectory
+(``all_latents``, ``all_log_probs``, ``all_timesteps``). All model-specific
+conditioning (text/image embeddings, condition latents, guidance state) is
+carried opaquely in ``condition`` and interpreted by each backend adapter, so
+the interface does not assume a particular text-encoder, CFG scheme, or
+condition-latent layout.
```
Comment (zhtmike) [inline verl_omni/pipelines/image_edit_utils.py]: keep header clean, possibility drop it

**Comment by NancyFyong** [comment]
@SamitHuang @zhtmike Thank you for your suggestion. I think the image editing and image generation interfaces can be merged.

After comparing the current t2i path with Qwen-Image-Edit and FLUX2/Klein-style image-conditioned generation, I think t2i and i2i/image editing can share the same high-level request shape:

- `images=[]` for text-to-image
- `images=[image]` for single-image editing / image-conditioned generation
- `images=[image0, image1, ...]` for future multi-image conditioning

At the task-interface level, the main difference is whether condition images are present. However, the backend details are still model-specific. For example, Qwen-Image-Edit feeds the condition image into both the VL processor/prompt encoding path and the VAE latent path, while FLUX2/Klein uses text-only prompt encoding and carries condition images through image latents and image ids. So I think the common interface should keep the request simple, and keep backend-specific replay inputs inside `condition`.

I will simplify the API in this direction:

- use a more general request interface for image generation instead of an image-edit-only request;
- keep the common replay fields as `all_latents`, `all_log_probs`, and `all_timesteps`;
- document `condition` as model-specific replay inputs, such as prompt/image embeddings, masks, condition latents, image ids, or shape metadata;
- keep `extras` for optional backend metadata.

This keeps the PR focused while making the interface usable for both t2i and i2i/image-editing backends. I am also happy to discuss this direction with everyone in the community meeting.

## Additional note about `prepare_processor_files`

The separate processor preparation hook is for models whose `processor/` directory exists but does not contain all config files expected by the Hugging Face processor loader. In that case, `hf_processor()` may fail before the model-specific adapter can run. Qwen-Image-Edit is one example where the processor files may need model-scoped normalization before loading.

Keeping this as `prepare_processor_files()` lets each model adapter fix its own processor layout before `hf_processor()` is called, without adding a global monkey patch.

**Comment by zhtmike** [comment]
Thanks for the clear explanation

**Comment by zhtmike** [comment]
pls ping me once the PR is ready, thx!

**Comment by zhtmike** [comment]
Too much file change, consider to split into 3 sub-prs for easier review :)
1. General Interface PR
2. Qwen-Image Editing PR
3. A document PR after PR1/PR2 merge?

**Comment by NancyFyong** [comment]
> Too much file change, consider to split into 3 sub-prs for easier review :)
> 
> 1. General Interface PR
> 2. Qwen-Image Editing PR
> 3. A document PR after PR1/PR2 merge?

Perhaps we could move the documentation pull request to the general interface pull request?

**Comment by zhtmike** [comment]
There will some API change during your `Qwen-Image Editing PR` ? I think the document will be better after everything finalized?

**Comment by NancyFyong** [comment]
> There will some API change during your `Qwen-Image Editing PR` ? I think the document will be better after everything finalized?

Sure, we can agree on a common interface first, then add documentation. I'll divide it into two branches first.

**Comment by NancyFyong** [comment]
@zhtmike The branch has been rebuilt from the latest upstream main and split as requested. This PR now contains only the general interface and its CPU tests. The Qwen-Image-Edit implementation is isolated on `NancyFyong/pr/qwen-image-edit`; I will open that follow-up against `main` after this PR merges so its diff does not repeat the General changes. The long integration document and FLUX work are not included here.

File: verl_omni/trainer/main_diffusion.py [hunk 9]
```diff
@@ -30,6 +31,22 @@
 from verl_omni.utils.diffusion_attention import fallback_fa3_if_unavailable, validate_attention_consistency
 
 
+def _prepare_processor_files(local_path: str, model_config) -> str | None:
```
Comment (zhtmike) [inline verl_omni/trainer/main_diffusion.py]: please make this function inline

File: verl_omni/pipelines/utils.py [hunk 10]
```diff
@@ -22,7 +25,92 @@
 
 from verl_omni.workers.config import DiffusionModelConfig
 
-from .model_base import DiffusionModelBase
+from .model_base import DiffusionI2IModelBase, DiffusionModelBase
+
+
+@dataclass
+class ImageGenerationRequest:
```
Comment (zhtmike) [inline verl_omni/pipelines/utils.py:44]: may I know when will it be called? in the next PR?

Comment (NancyFyong) [reply verl_omni/pipelines/utils.py:44]: yes，it will be used in qwen-image-edit pr

File: verl_omni/pipelines/model_base.py [hunk 11]
```diff
@@ -223,8 +244,205 @@ def forward(
         reverse-sampling algorithms (FlowGRPO et al.). Model adapters only need to
         override this when prediction requires extra handling such as CFG, negative
         inputs, or output conversion.
+
+        When ``_target_seq_len`` is present in ``model_inputs`` (set by
```
Comment (zhtmike) [inline verl_omni/pipelines/model_base.py]: is `_target_seq_len` universal? if it is not, we can overwrite `forward` for each i2i model later

File: verl_omni/pipelines/model_base.py [hunk 12]
```diff
@@ -223,8 +244,205 @@ def forward(
...
+        if isinstance(value, list | tuple):
+            if len(value) == 0:
+                return default
+            values = [_to_scalar(item) for item in value]
+            first = values[0]
+            if any(item != first for item in values[1:]):
+                raise ValueError(f"I2I metadata {key!r} differs across the micro-batch: {values!r}")
+            return first
+
+        return _to_scalar(value)
```
Comment (zhtmike) [inline verl_omni/pipelines/model_base.py]: quite hard to read, possible to check if tensordict_utils has similar kind of functions

File: verl_omni/pipelines/utils.py [hunk 13]
```diff
@@ -22,7 +25,92 @@
...
+
+        return cls(
+            prompt=prompt,
+            images=images,
+            negative_prompt=request_payload.get("negative_prompt"),
+            metadata=metadata,
+        )
+
+
+__all__ = [
```
Comment (zhtmike) [inline verl_omni/pipelines/utils.py]: put `__all__` at the top of file

File: verl_omni/pipelines/utils.py [hunk 14]
```diff
@@ -22,7 +25,92 @@
...
+
+        image_candidates = [
+            request_payload.get("images"),
+            request_payload.get("image"),
+            multi_modal_data.get("image") if isinstance(multi_modal_data, Mapping) else None,
+            extra_multi_modal_data.get("image") if isinstance(extra_multi_modal_data, Mapping) else None,
+            additional_information.get("condition_images") if isinstance(additional_information, Mapping) else None,
+        ]
+        images = []
+        for candidate in image_candidates:
```
Comment (zhtmike) [inline verl_omni/pipelines/utils.py:86]: write a one-line comment before the for-loop to show what it is doing

File: verl_omni/pipelines/utils.py [hunk 15]
```diff
@@ -22,7 +25,92 @@
...
+            images = list(images)
+        elif not isinstance(images, list):
+            images = [images]
+
+        metadata_candidates = [
+            request_payload.get("metadata"),
+            request_payload.get("extra_info"),
+            additional_information,
+        ]
+        metadata = None
```
Comment (zhtmike) [inline verl_omni/pipelines/utils.py:100]: similar, better to write a comment

File: verl_omni/pipelines/model_base.py [hunk 16]
```diff
@@ -223,10 +243,178 @@ def forward(
...
+                [
+                    inputs["hidden_states"],
+                    image_latents.to(
+                        device=inputs["hidden_states"].device,
+                        dtype=inputs["hidden_states"].dtype,
+                    ),
+                ],
+                dim=1,
+            )
+            inputs["_target_seq_len"] = target_seq_len
```
Comment (zhtmike) [inline verl_omni/pipelines/model_base.py]: I am thinking the SP related handling should be done outside of model base. Since it is non-trival case-by-case,  we can add SP hook later. WDYT

File: verl_omni/pipelines/model_base.py [hunk 17]
```diff
@@ -223,10 +243,178 @@ def forward(
         reverse-sampling algorithms (FlowGRPO et al.). Model adapters only need to
         override this when prediction requires extra handling such as CFG, negative
         inputs, or output conversion.
+
```
Comment (zhtmike) [inline verl_omni/pipelines/model_base.py]: extra blank

**Comment by zhtmike** [comment]
And also pls Fix CI & address AI review, thx

**Review by zhtmike — APPROVED** [review]
Thanks for the PR!
